### ISSUE TITLE
pyright flags as error alternative definitions of functions with the same name in if/then/else

### ISSUE THREAD

**Issue by fabiencro** [microsoft/pyright#495]
Hi,

Pyright seems to dislike it when a function has alternate definition within a if/then/else construct, one using a "def" statement and the other not.

See the following example: 
```python
import operator

use_complicated_key = True
mylist = []

if use_complicated_key:
    def sorting_key(item):
        return item[0]/item[1]+item[3]
else:
    sorting_key = operator.getitem(0)

mylist.sort(key=sorting_key)
```
pyright complains with `Declared function  already exists for 'sorting_key'`.

We also have the same problem when trying to assign `None` to a function name:
```python
if need_preprocessing:
    def preprocessing(string):
        return string +"_"
else:
    preprocessing = None
```

Interestingly, if we use `def` every time, it pyright accepts it:
```python
use_complicated_key = True
mylist = []

if use_complicated_key:
    def sorting_key(item):
        return item[0]/item[1]+item[3]
else:
    def sorting_key(item):
        return item[0]

mylist.sort(key=sorting_key)
```

**Comment by erictraut** [comment]
Although this is "legal" Python code, it's not a coding pattern I would recommend.

Pyright needs to make certain assumptions when it's performing static analysis. One of those assumptions is that declared function symbols won't be used only for declared functions, not reassigned to classes or other variables. The same is true for classes.

The recommended way to write the code above is to declare the function unconditionally and use a separate variable within the conditional statement. Not only will this work with Pyright, it will also be cleaner and easier-to-read code.

```
def _sort_key_complicated(item):
    return item[0]/item[1]+item[3]
if use_complicated_key:
   sorting_key = _sort_key_complicated
else:
    sorting_key = operator.getitem(0)
```

Or even better:
```
if use_complicated_key:
   sorting_key = lambda item: item[0]/item[1]+item[3]
else:
    sorting_key = operator.getitem(0)
```

**Comment by layday** [comment]
> The recommended way to write the code above is to declare the function unconditionally and use a separate variable within the conditional statement. Not only will this work with Pyright, it will also be cleaner and easier-to-read code.

I don't think this is markedly easier to read and begins to decrease in value as soon as you introduce additional conditions, like I have [here](https://github.com/layday/instawow/blob/0fae096/instawow/resolvers.py#L181-L189).

> `sorting_key = lambda item: item[0]/item[1]+item[3]`

This is discouraged by the Python style guide: "Always use a def statement instead of an assignment statement that binds a lambda expression directly to an identifier" [1].  I don't know, but I assume linters will mark it as an 'error'.

[1] https://www.python.org/dev/peps/pep-0008/

**Comment by erictraut** [comment]
I think Pyright is doing the right thing here. Here's my thinking.

I think we can all agree that the following code should produce a type error because the symbol `a` has been annotated with type `int`, and it's being assigned a value that is not compatible with `int`.
```
a: int = 3
a = 'hello'  # error: type mismatch
```

The same is true even if we put the two assignments in an if/else statement.
```
if value > 0:
   a: int = 3
else:
   a = 'hello'  # error: type mismatch
```

Python's type system doesn't provide a way to specify a type annotation for a function because the function declaration provides the type explicitly. It would be completely redundant to provide an annotation. The following is therefore also an error.
```
def my_function(p1: int, p2: str): pass
my_function = 'hello' # error: type mismatch
```

By analogy, the following is also an error:
```
if value > 0:
    def my_function(p1: int, p2: str) -> None: pass
else:
    my_function = None # error: type mismatch
```
